### PR TITLE
Add --clean flag

### PR DESCRIPTION
**Pull request by JakeSiFive**
This change adds the ability to clean all tracked outputs of jobs (except for source and claim jobs as determined by their command line).

Not all files created during a wake build are tracked by wake so this doesn't clean up *everything* but it should under normal uses clean up most things.

### CODE REVIEW

**Comment by JakeSiFive** [comment]
re: Wes on if nuking wake.db is useful....actually at first I wasn't going to delete it and then I sort of just decided to delete it anyway since the goal of a clean is to nuke as much as we can safely nuke I thought. I'm happy to go back since you pointed out that the `stats` table is still useful.

**Comment by JakeSiFive** [comment]
Another thing I haven't thought about, how does this interact with incremental builds invoked by wake?

**Comment by mmjconolly** [comment]
> Another thing I haven't thought about, how does this interact with incremental builds invoked by wake?

I think they will all be obliterated, given that it's a 'clean' function

File: tools/wake/main.cpp [hunk 1]
```diff
@@ -381,6 +383,31 @@ int main(int argc, char **argv) {
...
+
+    // Delete all the files
+    for (const auto& file : files) {
+      if (unlink(file.c_str()) == -1) {
+        if (errno == EISDIR) continue;
+        if (errno == ENOENT) continue;
+        std::cerr << "error: unlink(" << file << "): " << strerror(errno) << std::endl;
+        return 1;
+      }
+    }
```
Comment (mmjconolly) [inline tools/wake/main.cpp:400]: I'm wondering if we should (after file deletes) loop through the directories that wake created and delete any that are empty, leaving the non-empty ones as an obvious 'here is stuff here' indicator

**Comment by JakeSiFive** [comment]
> I think they will all be obliterated, given that it's a 'clean' function

That's what I *want* to happen but I'm not sure if we have incremental builds setup to be outputs. Wes showed me how that worked at one point but I've forgotten the details. The incremental inputs would have to be added as outputs in order for this to blow them away as written...I *think* that's how I would have implemented incremental builds but I could be wrong, I don't know.

File: tools/wake/main.cpp (hunk 1 above)
Comment (JakeSiFive) [reply tools/wake/main.cpp:400]: That does seem more tidy

**Comment by JakeSiFive** [comment]
I think I'm in favor of `git clean` instead honestly. Does anyone see an issue with people using `git clean`?

**Comment by mwachs5** [comment]
> Does anyone see an issue with people using git clean?

I don't usually expect "make clean" to delete the very files I may have been working on and forgotten to commit to git, perhaps being the reason I was confused and running `wake clean` in the first place trying to get stuff to work...

**Comment by mmjconolly** [comment]
Could be good to add 'clean' to `wake --help`

**Comment by mmjconolly** [comment]
I think we have another case to consider: `setPlanFnOutputs`

It seems like if a job writes some files and then filters the outputs, then they're remaining on the filesystem after a `--clean`.

```
$ cat do.wake 
package top
from wake import _
global def do _ =
    "touch foo; touch baa"
    | makePlan "" Nil
    | setPlanFnOutputs (filter (matches `foo` _))
    | runJob
    | getJobOutputs

$ ls foo baa
ls: cannot access 'foo': No such file or directory
ls: cannot access 'baa': No such file or directory
$ wake do
Pass (Path "foo", Nil)
$ ls foo baa
baa  foo
$ wake --clean
$ ls foo baa
ls: cannot access 'foo': No such file or directory
baa
```

**Comment by JakeSiFive** [comment]
> I think we have another case to consider: `setPlanFnOutputs`
> 
> If a job writes files and then filters the outputs, then those filtered-out are remaining on the filesystem after a `--clean`.
> 
> ```
> $ cat do.wake 
> package top
> from wake import _
> global def do _ =
>     "touch foo; touch baa"
>     | makePlan "" Nil
>     | setPlanFnOutputs (filter (matches `foo` _))
>     | runJob
>     | getJobOutputs
> 
> $ ls foo baa
> ls: cannot access 'foo': No such file or directory
> ls: cannot access 'baa': No such file or directory
> $ wake do
> Pass (Path "foo", Nil)
> $ ls foo baa
> baa  foo
> $ wake --clean
> $ ls foo baa
> ls: cannot access 'foo': No such file or directory
> baa
> ```
> 
> The usecases for I can think of are to avoid hashing large output files that we happen to know we don't care about, or to prevent visibility in a downstream job (though we could filter after getJobOutputs instead), or because multiple jobs produce the same file in a way that isn't of consequence (though I don't like that one)
> 
> Perhaps we could keep track of these files in the database too for the purposes of `--clean`? Though perhaps without hashing them?

I'm wondering if we should auto-delete such files after each job finishes? Or do we keep such files around for incremental builds?

**Comment by mmjconolly** [comment]
> > I think we have another case to consider: `setPlanFnOutputs`
> > If a job writes files and then filters the outputs, then those filtered-out are remaining on the filesystem after a `--clean`.
> > ```
> > $ cat do.wake 
> > package top
> > from wake import _
> > global def do _ =
> >     "touch foo; touch baa"
> >     | makePlan "" Nil
> >     | setPlanFnOutputs (filter (matches `foo` _))
> >     | runJob
> >     | getJobOutputs
> > 
> > $ ls foo baa
> > ls: cannot access 'foo': No such file or directory
> > ls: cannot access 'baa': No such file or directory
> > $ wake do
> > Pass (Path "foo", Nil)
> > $ ls foo baa
> > baa  foo
> > $ wake --clean
> > $ ls foo baa
> > ls: cannot access 'foo': No such file or directory
> > baa
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The usecases for I can think of are to avoid hashing large output files that we happen to know we don't care about, or to prevent visibility in a downstream job (though we could filter after getJobOutputs instead), or because multiple jobs produce the same file in a way that isn't of consequence (though I don't like that one)
> > Perhaps we could keep track of these files in the database too for the purposes of `--clean`? Though perhaps without hashing them?
> 
> I'm wondering if we should auto-delete such files after each job finishes? Or do we keep such files around for incremental builds?

I'm not opposed to the auto-delete, though also I don't know if there was a reason why we didn't initially do it

**Comment by mmjconolly** [comment]
@terpstra Do you know of any problems with auto-deleting filtered-out files when a job finishes?

**Comment by JakeSiFive** [comment]
Let's punt on this issue for now so we can get something that works well but perhaps isn't quite perfect landed. I think not cleaning up files removed by setPlanFnOutputs can be almost safely ignored.

File: src/runtime/database.cpp [hunk 2]
```diff
@@ -1091,6 +1126,20 @@ static std::vector<JobReflection> find_all(const Database *db, sqlite3_stmt *que
   return out;
 }
 
+std::vector<std::string> Database::get_outputs() const {
```
Comment (terpstra) [inline src/runtime/database.cpp:1129]: Maybe this should have been get_and_remove_outputs? So that the remove_output_files and get_output_files could be in the same begin/end_txn

File: test.cc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#define _XOPEN_SOURCE 700
```
Comment (terpstra) [inline test.cc:1]: Delete?

File: tools/fuse-waked/fuse-waked.cpp [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 // Open Group Base Specifications Issue 7
 #define _XOPEN_SOURCE 700
-#define FUSE_USE_VERSION 26
+#define FUSE_USE_VERSION 39
```
Comment (terpstra) [inline tools/fuse-waked/fuse-waked.cpp:23]: Maybe belongs in another PR?

File: tools/wake/main.cpp [hunk 5]
```diff
@@ -381,6 +388,64 @@ int main(int argc, char **argv) {
...
+  // If the user asked us to clean the local build, do so.
+  if (clean) {
+    // Find all the file we need to delete.
+    auto files = db.get_outputs();
+
+    // Clean up the database of unwanted info. Jobs must
+    // be cleared before outputs are removed to avoid foreign key
+    // constraint issues.
+    db.clear_jobs();
+    db.remove_outputs();
```
Comment (terpstra) [inline tools/wake/main.cpp:414]: All three calls into one txn, please.

File: tools/wake/main.cpp [hunk 6]
```diff
@@ -381,6 +388,64 @@ int main(int argc, char **argv) {
...
+          continue;
+        }
+        if (errno == ENOENT) continue;
+        std::cerr << "error: unlink(" << file << "): " << strerror(errno) << std::endl;
+        return 1;
+      }
+    }
+
+    // Sort them so that child directories come before parent directories
+    std::sort(dirs_to_remove.begin(), dirs_to_remove.end(), [&](const std::string& a, const std::string& b) -> bool {
```
Comment (terpstra) [inline tools/wake/main.cpp:433]: Just do your files erasure in the same pass post-sort.

**Review by terpstra — APPROVED** [review]
Easy fixes, but LGTM.
